src/main/java/org/candlepin/model/PoolQualifier.java

```diff
@@ -0,0 +1,532 @@
+     * satifies one of the provided patterns then the pool will be considered qualified for this condition.
+     * This pattern can include wildcards such as * and ?. Null and blank match values are filtered from the
+     * provided collection.
+     *
+     * @param matches
+     *  a collection of patterns for determining if a pool related field is considered qualified
+     *
+     * @return this instance
+     */
+    public PoolQualifier addMatches(Collection<String> matches) {
+        if (matches == null || matches.isEmpty()) {
+            return this;
+        }
+
+        matches.stream()
+            .filter(Objects::nonNull)
+            .filter(Predicate.not(String::isBlank))
+            .forEach(this.matches::add);
+
+        return this;
+    }
+
+    /**
+     * Removes a match pattern if it exists
+     *
+     * @param match
+     *  the match pattern to remove
+     *
+     * @return this instance
+     */
+    public PoolQualifier removeMatch(String match) {
+        if (match == null || match.isBlank()) {
+            return this;
+        }
+
+        this.matches.remove(match);
+        return this;
+    }
+
+    /**
+     * @return
+     *  an unmodifiable map of all the {@link Pool} product attributes that qualify a pool
+     */
+    public Map<String, List<String>> getAttributes() {
+        return Collections.unmodifiableMap(attributes);
+    }
+
+    /**
+     * Adds a {@link Pool} attribute condition that a pool must fulfill to be considered qualified. Multiple
+     * attribute values can be added for the same attribute key.
+     *
+     * @param key
+     *  the attribute key
+     *
+     * @param value
+     *  the attribute value
+     *
+     * @return this instance
+     */
+    public PoolQualifier addAttribute(String key, String value) {
+        if (key == null || key.isBlank()) {
+            return this;
+        }
+
+        attributes.computeIfAbsent(key, k -> new ArrayList<>())
+            .add(value);
+
+        return this;
+    }
+
+    /**
+     * Removes all {@link Pool} product attribute values based on a provided key.
+     *
+     * @param key
+     *  the pool attribute key to remove
+     *
+     * @return this instance
+     */
+    public PoolQualifier removeAttribute(String key) {
+        if (key == null || key.isBlank()) {
+            return this;
+        }
+
+        attributes.remove(key);
+        return this;
+    }
+
+    /**
+     * @return the date the pools must be active on to be considered qualified
+     */
+    public Date getActiveOn() {
+        return activeOn;
+    }
+
+    /**
+     * Sets a date that a {@link Pool} must have been active on based on the start and end date of the pool.
+     *
+     * @param activeOn
+     *  the date the pool must be active on to be considered qualified
+     *
+     * @return this instance
+     */
+    public PoolQualifier setActiveOn(Date activeOn) {
+        this.activeOn = activeOn;
```

Comment:
The Date class is mutable, so it may be a good idea to clone it into a new Date object before setting the local variable, to avoid bugs where its value is changed between when it is set, and when its read.
Same thing applies to other setters that accept Date in this class

Reply:
Updated throughout